Implement a toolchain

jadenPete opened the pull request:
This PR leverages [the Bazel toolchain framework](https://bazel.build/extending/toolchains) to replace `configure_bootstrap_scala`, `configure_zinc_scala`, and the `scala` attribute. Those rules were written before the framework existed, so it makes sense to replace it with a more modern, idiomatic alternative.

Note that this PR breaks existing repositories that use `lucidsoftware/rules_scala`, so it's necessary to migrate your repository before pulling in these changes. Here are the steps for doing so:
1. Remove the `default_scala` mapping in your `WORKSPACE` file
2. Replace calls to `configure_bootstrap_scala`, `configure_zinc_scala` with their `register_*_toolchain` equivalents
3. Update your call to `scala_register_toolchains` to provide a list of toolchains and a default toolchain
4. Profit

Commit:
Wrote bootstrap and Zinc Scala toolchains

Commit:
Lean more on ScalaRulePhase for phase configuration

Commit:
Use the configuration toolchains in the scala* rules

Commit:
Added transitions for the configuration toolchains

Commit:
Deleted tests/providers

It doesn't seem to test anything meaningful and hasn't been running
because the entire test is commented out.

Commit:
Updated the bazelbuild/rules_scala compatibility rules

Commit:
Removed reconfigure_deps_configuration

Commit:
Updated the usages of configure_bootstrap_scala and configure_zinc_scala

Commit:
Made the default toolchain configurable

Commit:
Removed configure_bootstrap_scala and configure_zinc_sala

Commit:
Updated the documentation accordingly

Removed an outdated comment about generating documentation, since
`scripts/gen-docs.sh` no longer exists. I also removed
docs/configure_zinc_scala.md`, since the rest of our autogenerated
documentation has been deleted.

jadenPete commented:
Rebased off of `lucid-master`.

jadenPete commented:
Rebased off of `lucid-master`.

Commit:
Use an outgoing transition to ensure Scala targets' dependencies are built once

jjudd commented on CONTRIBUTING.md:
```diff
@@ -2,14 +2,6 @@
 
 Contributions should follow the [principals](../README.md#principals) of rules_scala_annex.
 
-## Documentation
```
I believe this will conflict with your other PR to bring back doc.

jadenPete replied:
Yup, I'll need to re-add this section. Thanks for noticing that.

jjudd commented on rules/private/phases/api.bzl:
```diff
@@ -5,44 +5,56 @@ load(
+            [phase for phase_provider in phase_providers for phase in phase_provider.phases],
+        )
 
-    gd = {
```
Thanks for giving these more descriptive names

jadenPete replied:
No problem! Single letter or heavily abbreviated variable names tend to crawl under my skin (with the exception of `i`, `j`, `k`, etc.).

jjudd commented on rules/private/phases/phase_coverage_jacoco.bzl:
```diff
@@ -15,14 +15,9 @@ def phase_coverage_jacoco(ctx, g):
     if not ctx.configuration.coverage_enabled:
         return
 
-    if _CodeCoverageConfiguration not in ctx.attr.scala:
```
Is this early exit not longer relevant because the toolchain will always have the code coverage configuration?

jadenPete replied:
Not quite. This change is actually unrelated to toolchains. Previously, we included this phase in both the bootstrap and Zinc rules, despite this early exit preventing it from doing anything in the bootstrap rules.

I thought that was dumb, so I enabled the phase only for the Zinc rules and removed this statement.

jjudd commented on rules/private/phases/phase_semanticdb.bzl:
```diff
@@ -8,9 +8,9 @@ load("@rules_scala_annex//rules:providers.bzl", _ScalaConfiguration = "ScalaConf
 #
 def phase_semanticdb(ctx, g):
-    scala_configuration = ctx.attr.scala[_ScalaConfiguration]
+    toolchain = ctx.toolchains["//rules/scala:toolchain_type"]
```
I imagine this string will never change, but it may be worth doing what the `rules_go` folks did and instead use a constant that we import. That way if this string ever needs to change, it will be easy to change.

https://github.com/bazel-contrib/rules_go/blob/a54de7c88bb3cb133f951f0e9817c09be42056cb/go/private/common.bzl#L16

Also, they use the fully qualified name, which is probably  not a bad idea either. So `@rules_scala_annex//rules/scala:toolchain_type`

jadenPete replied:
This sounds like a great idea, but if we ever need to move that constant, we'll need to update all the imports that reference it, which introduces the same problem.

Also, Is the fully-qualified name unnecessary? I prefer not to use those unless it's necessary. My understanding was that Bazel understood this was in `rules_scala_annex`, even if these rules are being used in another repository.

jadenPete replied:
Oh, you meant defining that alias for internal use in this file. That makes a lot of sense; I'll do that.

jjudd commented on rules/private/phases/phase_zinc_depscheck.bzl:
```diff
@@ -16,10 +16,7 @@ load(
 #
 
 def phase_zinc_depscheck(ctx, g):
-    if _DepsConfiguration not in ctx.attr.scala:
```
Similar question here: is the early exit no longer needed because the toolchain will always have the deps configuration?

jadenPete replied:
To be honest, I don't understand why this statement was here in the first place. Not only does the toolchain always have this provider, but `ctx.attr.scala` should've as well.

jjudd commented on rules/register_toolchain.bzl:
```diff
@@ -0,0 +1,238 @@
+register_zinc_toolchain = _make_register_toolchain(_zinc_configuration)
+
+def _scala_toolchain_incoming_transition_impl(settings, attr):
+    if attr.scala_toolchain_name == "":
```
Should this be something like 
`if attr.scala_toolchain_name == "" or attr.scala_toolchain_name == settings["@rules_scala_annex_scala_toolchain//:scala-toolchain"]`

That way you don't transition if it's empty or if it's what is currently is.

jadenPete replied:
I think the check you're describing was rendered unnecessary by Bazel 7.0, which is exactly what makes the outgoing transition in this PR work:
https://github.com/bazelbuild/bazel/pull/16910

Basically, Bazel caches execution (and analysis?) according to the hash of the difference between the current configuration and the "baseline" (that specified in `.bazelrc` and on the command line), instead of re-running those phases every time we transition.

Thanks for thinking of that, though!

jjudd commented on rules/register_toolchain.bzl:
```diff
@@ -0,0 +1,238 @@
+    if attr.scala_toolchain_name == "":
+        return {}
+
+    return {
```
Do we want to be storing the original scala_toolchain_name like the `rules_go` folks did?

Because currently this will always reset to the original value, and I thought that you ran into a scenario where you could override things from the command line, which made that problematic.

I did this in the Play Routes and Twirl rules, in case it is at all helpful
```python
    new_settings = {}
    new_settings[toolchain_setting_key] = attr.play_routes_toolchain_name

    # Store the original toolchain value, so we can reset it
    new_settings[original_toolchain_setting_key] = json.encode(settings[toolchain_setting_key])

    return new_settings
```

Then I reset it as follows:
```python
    if settings[original_toolchain_setting_key] != "":
        new_settings = {}
        new_settings[toolchain_setting_key] = json.decode(settings[original_toolchain_setting_key])
        new_settings[original_toolchain_setting_key] = ""
        return new_settings
    else:
        return {}
```

jadenPete replied:
Ah, I completely forgot about that. I'll go ahead and do that.

jjudd commented on rules/rules_scala.bzl:
```diff
@@ -56,12 +56,7 @@ emulate_rules_scala_repository = repository_rule(
 )
 
-def emulate_rules_scala(scala, scalatest, extra_deps = []):
-    native.bind(
```
Do we need to replace this with anything? I imagine this is because the version of rules_scala from the compat layer is so old it's still using bind? I imagine it's not worth upgrading at this point, but I figured I'd ask.

jadenPete replied:
Nope! That was used here, but it's no longer used:
https://github.com/lucidsoftware/rules_scala/pull/51/files#diff-471416c47abcd84fb475b6b41ec3ed263aff9093ffd10f9f807fb9e38ddb09acL26

jjudd commented on tests/WORKSPACE:
```diff
@@ -93,7 +93,24 @@ load("@annex//:defs.bzl", annex_pinned_maven_install = "pinned_maven_install")
+scala_register_toolchains(
+    default_scala_toolchain_name = "test_zinc_2_13",
+    toolchains = [
+        "@@//dependencies/indirect:test_zinc_2_13_direct_deps_off",
```
Do we need the `@@` for these or can we just use the single `@`?

jadenPete replied:
You're right—it can. Do you know the difference between `@//...` and `@@//...`? I know the difference between `@/foo//...` and `@@foo//...`, but are  `@//...` and `@@//...` the same?

jjudd replied:
@@ is canonical name. Was introduced with bzlmod. AFAIK it's intended to be machine readable rather than human readable. https://bazel.build/rules/lib/builtins/Args.html

> [Label](https://bazel.build/rules/lib/builtins/Label.html) objects are turned into a string representation that resolves back to the same object when resolved in the context of the main repository. If possible, the string representation uses the apparent name of a repository in favor of the repository's canonical name, which makes this representation suited for use in BUILD files. While the exact form of the representation is not guaranteed, typical examples are //foo:bar, @repo//foo:bar and @@canonical_name+//foo:bar.bzl.

jjudd replied:
Also https://bazel.build/concepts/labels

This link explains it better.

jjudd commented on tests/scala/toolchain/Parent.scala:
```diff
@@ -0,0 +1,5 @@
+object Parent {
```
Would it be worth using Scala 3 specific features in this file, so it breaks if somehow you end up with a Scala 2 toolchain here instead of 3?

jadenPete replied:
I could, but I don't think that would cause it to break because this test only performs a `bazel cquery`, not a `bazel build`.

jjudd commented on tests/scala/toolchain/test:
```diff
@@ -0,0 +1,18 @@
+
+bazel_cquery_output="$(
+    bazel cquery 'deps(:parent1 + :parent2, 1)' --output graph --nograph:factored |&
+        sed -nr 's/^ *"\/\/scala\/toolchain:child \(([0-9a-f]*)\)"$/\1/p'
```
It would be helpful if you could add a comment here describing what you're checking for, so folks don't have to spend the mental energy to decode the sed command unless they really need to.

jadenPete replied:
Good idea :sweat_smile:

jjudd reviewed (commented):
Took a pass through and things mostly looked good to me. I had a few questions/comments.

jjudd commented:
One thing that's been bouncing around in my head and I don't know if I care about or not: we use `scala_toolchain_name`. I used a similar naming convention in the Play Routes and Twirl rules. 

I wonder if some folks will think that means they're getting that particular label. When in reality they should get a toolchain with that name, which works with the right platform? 

I don't think `scala_toolchain_kind` or `scala_toolchain_type` would necessarily be clearer. Curious your thoughts here.

jadenPete commented:
Squashed.

Commit:
Fixed various failing tests

These tests were looking in the wrong output directory, which has
changed now that we're using a configuration setting to determine the
Scala compiler.

jadenPete commented:
Rebased off of `lucid-master`.

jadenPete commented:
> One thing that's been bouncing around in my head and I don't know if I care about or not: we use `scala_toolchain_name`. I used a similar naming convention in the Play Routes and Twirl rules.
> 
> I wonder if some folks will think that means they're getting that particular label. When in reality they should get a toolchain with that name, which works with the right platform?
> 
> I don't think `scala_toolchain_kind` or `scala_toolchain_type` would necessarily be clearer. Curious your thoughts here.

Yeah, I remember deliberating about the naming of `scala_toolchain_name` for some time. The attribute that's used in `bazelbuild/rules_scala` stores the desired Scala version, so I couldn't refer to that. I think "name" is clear enough, because if `scala_toolchain_name` stored the label of the desired toolchain, its name would be `scala_toolchain_label`, or just `scala_toolchain`.

Commit:
Disabled tests/dependencies/indirect/test

jadenPete commented:
Squashed.

jadenPete commented:
Squashed.
